docs/user_utilities.rst

```diff
@@ -0,0 +1,154 @@
+
+    :parameter info: (string) logging info
+
+    Print out the input string in blue color to stdout, only when the logging level is equal to ``ti.TRACE``.
+
+.. function:: ti.error(info)
+
+    :parameter info: (string) logging info
+
+    This function prints out the input string in red color to stdout in any logging level.
+
+    Note that it crashes the Taichi kernel as well as your Taichi program then throw an ``RuntimeError`` out.
+
+.. Warn::
+
+    The usages of ``ti.trace()``, ``ti.debug()`` and ``ti.warn()`` are all similar to ``ti.info()``.But ``ti.error()`` is different from them. When it is called, the program will be crashed immediately.
+
+    In spite of this, the output of ``ti.error()`` is still ``stdout`` but not ``stderr``.
+
+Here is an example:
+
+.. code-block:: python
+
+    import taichi as ti
+
+    ti.init()
+    ti.set_logging_level(ti.ERROR)
+
+    try:
+        ti.error("It's a big error! Now exit...")
+    except RuntimeError as err:
+        print(err)
+
+Profiler
+--------
+
+Taichi profiler can help you count the time cost of your program. There are 2 profilers in Taichi available at this moment: ``ScopedProfiler`` and ``ProfilerBase``.
+
+ScopedProfiler
+##############
+
+1. ``ScopedProfiler`` evaluates the time spent for each step of the **host tasks** during execution hierarchically.
+
+2. Moreover, this profiler is turned on automatically and no extra work is needed except writing ``ti.core.print_profile_info()`` at the end of your Taichi program. For example:
+
+.. code-block:: python
+
+    ti.init(arch = ti.cpu)
+    var = ti.var(ti.f32, shape = 1)
+
+    @ti.kernel
+    def compute():
+        var[0] = 1.0
+        print(f" set var[0] = {var[0]}")
+
+    compute()
+    ti.core.print_profile_info()
+
+``ti.core.print_profile_info()`` will output statistics in a hierarchical format, with different colors for different levels.
+
+.. Note::
+
+    ``ScopedProfiler`` is a c++ class inside of Taichi core, so you cannot see it directly in python scope.
+
+    We mention its name here is just to make a distinction with another profiler.
+
+ProfilerBase
+############
+
+1. ``ProfilerBase`` records the cost of kernel launches on devices. ``kernel_profiler`` must be set to ``True`` in ``ti.init`` to enable this profiler.
+
+2. When all the computation is finished in your program, call ``ti.kernel_profiler_print()`` to show to profiling result. For example:
+
+.. code-block :: python
+    :emphasize-lines: 1, 9
+
+    ti.init(ti.cpu, kernel_profiler = True)
+    var = ti.var(ti.f32, shape = 1)
+
+    @ti.kernel
+    def compute():
+        var[0] = 1.0
+
+    compute()
+    ti.kernel_profiler_print()
+
+Its output is like:
+
+::
+
+    [ 22.73%] jit_evaluator_0_kernel_0_serial             min   0.001 ms   avg   0.001 ms   max   0.001 ms   total   0.000 s [      1x]
+    [  0.00%] jit_evaluator_1_kernel_1_serial             min   0.000 ms   avg   0.000 ms   max   0.000 ms   total   0.000 s [      1x]
```

Comment:
OFT: Can we make jit&accessor kernels invisible by default? They are usually 0 and confusing to end-users.